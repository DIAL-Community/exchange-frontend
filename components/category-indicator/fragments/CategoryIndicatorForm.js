import React, { useState, useCallback, useContext } from 'react'
import { useRouter } from 'next/router'
import { useMutation } from '@apollo/client'
import { useIntl } from 'react-intl'
import { FaSpinner } from 'react-icons/fa'
import { Controller, useForm } from 'react-hook-form'
import Input from '../../shared/form/Input'
import Select from '../../shared/form/Select'
import { useUser } from '../../../lib/hooks'
import { HtmlEditor } from '../../shared/form/HtmlEditor'
import { Loading, Unauthorized } from '../../shared/FetchStatus'
import ValidationError from '../../shared/form/ValidationError'
import { ToastContext } from '../../../lib/ToastContext'
import { generateCategoryIndicatorTypes } from '../../shared/form/options'
import { CREATE_CATEGORY_INDICATOR } from '../../shared/mutation/categoryIndicator'

const CategoryIndicatorForm = ({ rubricCategory, categoryIndicator }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const { showSuccessMessage, showFailureMessage } = useContext(ToastContext)

  const slug = categoryIndicator?.slug ?? ''

  const router = useRouter()
  const { locale } = router

  const { user, isAdminUser, isEditorUser, loadingUserSession } = useUser()

  const [mutating, setMutating] = useState(false)
  const [reverting, setReverting] = useState(false)

  const categoryIndicatorTypes = generateCategoryIndicatorTypes(format)

  const [updateCategoryIndicator, { reset }] = useMutation(CREATE_CATEGORY_INDICATOR, {
    onCompleted: (data) => {
      const { createCategoryIndicator: response } = data
      if (response?.categoryIndicator && response?.errors?.length === 0) {
        setMutating(false)
        showSuccessMessage(
          format('toast.submit.success', { entity: format('categoryIndicator.label') }),
          () => router.push(
            `/${locale}` +
            `/rubric-categories/${rubricCategory.slug}` +
            `/category-indicators/${response.categoryIndicator.slug}`
          )
        )
      } else {
        showFailureMessage(format('toast.submit.failure', { entity: format('categoryIndicator.label') }))
        setMutating(false)
        reset()
      }
    },
    onError: () => {
      showFailureMessage(format('toast.submit.failure', { entity: format('categoryIndicator.label') }))
      setMutating(false)
      reset()
    }
  })

  const { handleSubmit, register, control, formState: { errors } } = useForm({
    mode: 'onSubmit',
    reValidateMode: 'onChange',
    shouldUnregister: true,
    defaultValues: {
      name: categoryIndicator?.name,
      weight: categoryIndicator?.weight,
      indicatorType: categoryIndicatorTypes.find(
        ({ value: indicatorType }) => indicatorType === categoryIndicator?.indicatorType
      ),
      dataSource: categoryIndicator?.dataSource,
      scriptName: categoryIndicator?.scriptName,
      description: categoryIndicator?.categoryIndicatorDescription?.description
    }
  })

  const doUpsert = async (data) => {
    if (user) {
      setMutating(true)
      const { userEmail, userToken } = user
      const { name, description, dataSource, scriptName, indicatorType } = data
      const weight = parseFloat(data.weight)
      const rubricCategorySlug = rubricCategory?.slug

      const variables = {
        name,
        slug,
        rubricCategorySlug,
        indicatorType: indicatorType.value,
        weight,
        dataSource,
        scriptName,
        description
      }

      updateCategoryIndicator({
        variables,
        context: {
          headers: {
            'Accept-Language': locale,
            Authorization: `${userEmail} ${userToken}`
          }
        }
      })
    }
  }

  const cancelForm = () => {
    setReverting(true)

    router.push(
      `/rubric-categories/${rubricCategory?.slug}`+
      `/category-indicators/${categoryIndicator?.slug ?? ''}`
    )
  }

  return loadingUserSession
    ? <Loading />
    : isAdminUser || isEditorUser
      ? (
        <form onSubmit={handleSubmit(doUpsert)}>
          <div className='px-4 lg:px-0 py-4 lg:py-6 text-dial-plum'>
            <div className='flex flex-col gap-y-6 text-sm'>
              <div className='text-xl font-semibold'>
                {categoryIndicator
                  ? format('app.editEntity', { entity: categoryIndicator.name })
                  : `${format('app.createNew')} ${format('categoryIndicator.label')}`
                }
              </div>
              <div className='flex flex-col gap-y-2'>
                <label className='text-dial-sapphire required-field' htmlFor='name'>
                  {format('app.name')}
                </label>
                <Input
                  {...register('name', { required: format('validation.required') })}
                  id='name'
                  placeholder={format('app.name')}
                  isInvalid={errors.name}
                />
                {errors.name && <ValidationError value={errors.name?.message} />}
              </div>
              <div className='flex flex-col gap-y-2'>
                <label className='text-dial-sapphire required-field'>
                  {format('categoryIndicator.indicatorType')}
                </label>
                <Controller
                  name='indicatorType'
                  control={control}
                  render={({ field }) => (
                    <Select
                      {...field}
                      isSearch
                      options={categoryIndicatorTypes}
                      placeholder={format('categoryIndicator.indicatorType')}
                      isInvalid={errors.indicatorType}
                    />
                  )}
                  rules={{ required: format('validation.required') }}
                />
                {errors.indicatorType && <ValidationError value={errors.indicatorType?.message} />}
              </div>
              <div className='flex flex-col gap-y-2'>
                <label className='text-dial-sapphire required-field' htmlFor='weight'>
                  {format('categoryIndicator.weight')}
                </label>
                <Input
                  {...register('weight', { required: format('validation.required') })}
                  id='weight'
                  type='number'
                  step='0.001'
                  placeholder={format('categoryIndicator.weight')}
                  isInvalid={errors.weight}
                />
                {errors.weight && <ValidationError value={errors.weight?.message} />}
              </div>
              <div className='form-field-wrapper'>
                <label className='text-dial-sapphire '>
                  {format('categoryIndicator.dataSource')}
                </label>
                <Input {...register('dataSource')} placeholder={format('categoryIndicator.dataSource')} />
              </div>
              <div className='form-field-wrapper'>
                <label className='text-dial-sapphire '>
                  {format('categoryIndicator.scriptName')}
                </label>
                <Input {...register('scriptName')} placeholder={format('categoryIndicator.scriptName')} />
              </div>
              <div className='flex flex-col gap-y-2'>
                <label className='text-dial-sapphire required-field'>
                  {format('app.description')}
                </label>
                <Controller
                  name='description'
                  control={control}
                  render={({ field: { value, onChange } }) => (
                    <HtmlEditor
                      editorId='description-editor'
                      onChange={onChange}
                      initialContent={value}
                      placeholder={format('app.description')}
                      isInvalid={errors.description}
                    />
                  )}
                  rules={{ required: format('validation.required') }}
                />
                {errors.description && <ValidationError value={errors.description?.message} />}
              </div>
              <div className='flex flex-wrap text-base mt-6 gap-3'>
                <button
                  type='submit'
                  className='submit-button'
                  disabled={mutating || reverting}
                >
                  {`${format('app.submit')} ${format('categoryIndicator.label')}`}
                  {mutating && <FaSpinner className='spinner ml-3' />}
                </button>
                <button
                  type='button'
                  className='cancel-button'
                  disabled={mutating || reverting}
                  onClick={cancelForm}
                >
                  {format('app.cancel')}
                  {reverting && <FaSpinner className='spinner ml-3' />}
                </button>
              </div>
            </div>
          </div>
        </form>
      )
      : <Unauthorized />
}

CategoryIndicatorForm.displayName = 'CategoryIndicatorForm'

export default CategoryIndicatorForm
