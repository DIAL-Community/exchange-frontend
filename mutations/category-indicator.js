import { gql } from '@apollo/client'

export const CREATE_CATEGORY_INDICATOR = gql`
  mutation CreateCategoryIndicator (
    $name: String!
    $slug: String!
    $rubricCategorySlug: String
    $indicatorType: String!
    $dataSource: String
    $scriptName: String
    $weight: Float!
    $description: String!
  ) {
    createCategoryIndicator(
      name: $name
      slug: $slug
      rubricCategorySlug: $rubricCategorySlug
      indicatorType: $indicatorType
      dataSource: $dataSource
      scriptName: $scriptName
      weight: $weight
      description: $description
    ) {
      categoryIndicator{
        id
        slug
      }
      errors
    }
  }
`

export const DELETE_CATEGORY_INDICATOR = gql`
  mutation DeleteCategoryIndicator($id: ID!) {
    deleteCategoryIndicator(id: $id) {
      rubricCategorySlug
    }
  }
`
