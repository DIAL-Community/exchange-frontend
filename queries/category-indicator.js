import { gql } from '@apollo/client'

export const CATEGORY_INDICATOR_QUERY = gql`
  query CategoryIndicator($slug: String!) {
    categoryIndicator(slug: $slug) {
      slug
      name
      id
      indicatorType
      weight
      dataSource
      scriptName
      rubricCategoryId
      categoryIndicatorDescription {
        description
      } 
    }
  }
`
