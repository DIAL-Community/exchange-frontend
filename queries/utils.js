const MIN_SEARCH_VALUE_CHARACTERS = 2

export const fetchSelectOptions = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}

export const fetchSelectOptionsWithMature = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input,
    mature: true
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}
