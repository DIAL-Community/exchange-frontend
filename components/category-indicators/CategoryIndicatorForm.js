import React, { useState, useCallback, useMemo, useContext } from 'react'
import { useRouter } from 'next/router'
import { useMutation } from '@apollo/client'
import { useIntl } from 'react-intl'
import { FaSpinner } from 'react-icons/fa'
import { Controller, useForm } from 'react-hook-form'
import Breadcrumb from '../shared/breadcrumb'
import { HtmlEditor } from '../shared/HtmlEditor'
import Input from '../shared/Input'
import { ToastContext } from '../../lib/ToastContext'
import ValidationError from '../shared/ValidationError'
import { useUser } from '../../lib/hooks'
import { getCategoryIndicatorTypes } from '../../lib/utilities'
import Select from '../shared/Select'
import { CREATE_CATEGORY_INDICATOR } from '../../mutations/category-indicator'

const CategoryIndicatorForm = ({ rubricCategory, categoryIndicator }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const { showToast } = useContext(ToastContext)

  const slug = categoryIndicator?.slug ?? ''

  const router = useRouter()
  const { locale } = router

  const { user } = useUser()

  const [mutating, setMutating] = useState(false)
  const [reverting, setReverting] = useState(false)

  const categoryIndicatorTypes = getCategoryIndicatorTypes(format)

  const [updateCategoryIndicator, { reset }] = useMutation(CREATE_CATEGORY_INDICATOR, {
    onCompleted: (data) => {
      const { createCategoryIndicator: response } = data
      if (response?.categoryIndicator && response?.errors?.length === 0) {
        setMutating(false)
        showToast(
          format('toast.category-indicator.submit.success'),
          'success',
          'top-center',
          1000,
          null,
          () => router.push(
            `/${locale}` +
            `/rubric_categories/${rubricCategory.slug}` +
            `/${data.createCategoryIndicator.categoryIndicator.slug}`
          )
        )
      } else {
        setMutating(false)
        showToast(format('toast.category-indicator.submit.failure'), 'error', 'top-center')
        reset()
      }
    },
    onError: (error) => {
      setMutating(false)
      showToast(
        <div className='flex flex-col'>
          <span>{format('toast.category-indicator.submit.failure')}</span>
          <span>{error?.message}</span>
        </div>,
        'error',
        'top-center'
      )
      reset()
    }
  })

  const { handleSubmit, register, control, formState: { errors } } = useForm({
    mode: 'onSubmit',
    reValidateMode: 'onChange',
    shouldUnregister: true,
    defaultValues: {
      name: categoryIndicator?.name,
      weight: categoryIndicator?.weight,
      indicatorType: categoryIndicatorTypes.find(
        ({ value: indicatorType }) => indicatorType === categoryIndicator?.indicatorType
      ),
      dataSource: categoryIndicator?.dataSource,
      scriptName: categoryIndicator?.scriptName,
      description: categoryIndicator?.categoryIndicatorDescription?.description
    }
  })

  const slugNameMapping = useMemo(() => {
    const map = {
      create_category_indicator: format('categoryIndicator.create'),
      edit: format('app.edit')
    }

    if (rubricCategory) {
      map[rubricCategory.slug] = rubricCategory.name
    }

    if (categoryIndicator) {
      map[categoryIndicator.slug] = categoryIndicator.name
    }

    return map
  }, [categoryIndicator, rubricCategory, format])

  const doUpsert = async (data) => {
    if (user) {
      setMutating(true)
      const { userEmail, userToken } = user
      const { name, description, dataSource, scriptName, indicatorType } = data
      const weight = parseFloat(data.weight)
      const rubricCategorySlug = rubricCategory?.slug

      const variables = {
        name,
        slug,
        rubricCategorySlug,
        indicatorType: indicatorType.value,
        weight,
        dataSource,
        scriptName,
        description
      }

      updateCategoryIndicator({
        variables,
        context: {
          headers: {
            'Accept-Language': locale,
            Authorization: `${userEmail} ${userToken}`
          }
        }
      })
    }
  }

  const cancelForm = () => {
    setReverting(true)

    router.push(`/rubric_categories/${rubricCategory?.slug}/${categoryIndicator?.slug ?? ''}`)
  }

  return (
    <div className='flex flex-col'>
      <div className='hidden lg:block px-8'>
        <Breadcrumb slugNameMapping={slugNameMapping} />
      </div>
      <div className='pb-8 px-8'>
        <div id='content' className='sm:px-0 max-w-full mx-auto'>
          <form onSubmit={handleSubmit(doUpsert)}>
            <div className='bg-edit shadow-md rounded px-8 pt-6 pb-12 mb-4 flex flex-col gap-3'>
              <div className='text-2xl font-bold text-dial-blue pb-4'>
                {categoryIndicator
                  ? format('app.edit-entity', { entity: categoryIndicator.name })
                  : `${format('app.create-new')} ${format('categoryIndicator.label')}`
                }
              </div>
              <div className='flex flex-col lg:flex-row gap-4'>
                <div className='w-full lg:w-1/2 flex flex-col gap-y-3'>
                  <div className='form-field-wrapper' data-testid='category-indicator-name'>
                    <label className='form-field-label required-field' htmlFor='name'>
                      {format('app.name')}
                    </label>
                    <Input
                      {...register('name', { required: format('validation.required') })}
                      id='name'
                      placeholder={format('app.name')}
                      isInvalid={errors.name}
                    />
                    {errors.name && <ValidationError value={errors.name?.message} />}
                  </div>
                  <div className='form-field-wrapper' data-testid='category-indicator-indicator-type'>
                    <label className='form-field-label required-field'>
                      {format('categoryIndicator.indicatorType')}
                    </label>
                    <Controller
                      name='indicatorType'
                      control={control}
                      render={({ field }) => (
                        <Select
                          {...field}
                          isSearch
                          options={categoryIndicatorTypes}
                          placeholder={format('categoryIndicator.indicatorType')}
                          isInvalid={errors.indicatorType}
                        />
                      )}
                      rules={{ required: format('validation.required') }}
                    />
                    {errors.indicatorType && <ValidationError value={errors.indicatorType?.message} />}
                  </div>
                  <div className='form-field-wrapper' data-testid='category-indicator-weight'>
                    <label className='form-field-label required-field' htmlFor='weight'>
                      {format('categoryIndicator.weight')}
                    </label>
                    <Input
                      {...register('weight', { required: format('validation.required') })}
                      id='weight'
                      type='number'
                      step='0.001'
                      placeholder={format('categoryIndicator.weight')}
                      isInvalid={errors.weight}
                    />
                    {errors.weight && <ValidationError value={errors.weight?.message} />}
                  </div>
                  <div className='form-field-wrapper'>
                    <label className='form-field-label'>
                      {format('categoryIndicator.dataSource')}
                    </label>
                    <Input {...register('dataSource')} placeholder={format('categoryIndicator.dataSource')} />
                  </div>
                  <div className='form-field-wrapper'>
                    <label className='form-field-label'>
                      {format('categoryIndicator.scriptName')}
                    </label>
                    <Input {...register('scriptName')} placeholder={format('categoryIndicator.scriptName')} />
                  </div>
                </div>
                <div className='w-full lg:w-1/2'>
                  <div className='flex flex-col gap-y-2' data-testid='category-indicator-description'>
                    <label className='form-field-label required-field'>
                      {format('app.description')}
                    </label>
                    <Controller
                      name='description'
                      control={control}
                      render={({ field: { value, onChange } }) => (
                        <HtmlEditor
                          editorId='description-editor'
                          onChange={onChange}
                          initialContent={value}
                          placeholder={format('app.description')}
                          isInvalid={errors.description}
                        />
                      )}
                      rules={{ required: format('validation.required') }}
                    />
                    {errors.description && <ValidationError value={errors.description?.message} />}
                  </div>
                </div>
              </div>
              <div className='flex flex-wrap text-xl mt-8 gap-3'>
                <button
                  type='submit'
                  className='submit-button'
                  disabled={mutating || reverting}
                  data-testid='submit-button'
                >
                  {`${format('app.submit')} ${format('categoryIndicator.label')}`}
                  {mutating && <FaSpinner className='spinner ml-3' />}
                </button>
                <button
                  type='button'
                  className='cancel-button'
                  disabled={mutating || reverting}
                  onClick={cancelForm}
                >
                  {format('app.cancel')}
                  {reverting && <FaSpinner className='spinner ml-3' />}
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  )
}

CategoryIndicatorForm.displayName = 'CategoryIndicatorForm'

export default CategoryIndicatorForm
