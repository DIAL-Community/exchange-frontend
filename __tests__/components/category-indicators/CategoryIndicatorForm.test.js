import { fireEvent, screen, waitFor } from '@testing-library/react'
import { act } from 'react-dom/test-utils'
import userEvent from '@testing-library/user-event'
import CustomMockedProvider, { generateMockApolloData } from '../../utils/CustomMockedProvider'
import { render, waitForAllEffectsAndSelectToLoad } from '../../test-utils'
import { CREATE_CATEGORY_INDICATOR } from '../../../mutations/category-indicator'
import CategoryIndicatorForm from '../../../components/category-indicators/CategoryIndicatorForm'
import { mockNextAuthUseSession, mockNextUseRouter, statuses } from '../../utils/nextMockImplementation'
import {
  createCategoryIndicatorFailure,
  createCategoryIndicatorSuccess,
  categoryIndicator,
  rubricCategory
} from './data/CategoryIndicatorForm'

mockNextUseRouter()

describe('Unit tests for CategoryIndicatorForm component.', () => {
  const CATEGORY_INDICATOR_NAME_TEST_ID = 'category-indicator-name'
  const CATEGORY_INDICATOR_WEIGHT_TEST_ID = 'category-indicator-weight'
  const CATEGORY_INDICATOR_INDICATOR_TYPE_TEST_ID = 'category-indicator-indicator-type'
  const CATEGORY_INDICATOR_INDICATOR_TYPE_OPTION_LABEL = 'numeric'
  const CATEGORY_INDICATOR_DESCRIPTION_TEST_ID = 'category-indicator-description'
  const SUBMIT_BUTTON_TEST_ID = 'submit-button'
  const REQUIRED_FIELD_MESSAGE = 'This field is required'

  test('Should render CategoryIndicatorForm component for admin user.', async () => {
    mockNextAuthUseSession(statuses.AUTHENTICATED, { isAdminUser: true })
    const { container } = render(
      <CustomMockedProvider>
        <CategoryIndicatorForm rubricCategory={rubricCategory} />
      </CustomMockedProvider>
    )
    await waitForAllEffectsAndSelectToLoad(container)
    expect(container).toMatchSnapshot()
  })

  test('Should show validation errors for mandatory fields and hide them on input value change.', async () => {
    const user = userEvent.setup()
    mockNextAuthUseSession(statuses.AUTHENTICATED, { isAdminUser: true })
    const { getByTestId, getByText, container } = render(
      <CustomMockedProvider>
        <CategoryIndicatorForm rubricCategory={rubricCategory} />
      </CustomMockedProvider>
    )
    await waitForAllEffectsAndSelectToLoad(container)
    await act(() => fireEvent.submit(getByTestId(SUBMIT_BUTTON_TEST_ID)))
    expect(getByTestId(CATEGORY_INDICATOR_NAME_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CATEGORY_INDICATOR_DESCRIPTION_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CATEGORY_INDICATOR_WEIGHT_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CATEGORY_INDICATOR_INDICATOR_TYPE_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await user.type(screen.getByLabelText(/Name/), 'test category indicator name')
    expect(getByTestId(CATEGORY_INDICATOR_NAME_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    await user.clear(screen.getByLabelText(/Name/))
    expect(getByTestId(CATEGORY_INDICATOR_NAME_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await user.type(screen.getByLabelText(/Name/), 'test category indicator name 2')
    expect(getByTestId(CATEGORY_INDICATOR_NAME_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CATEGORY_INDICATOR_DESCRIPTION_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CATEGORY_INDICATOR_WEIGHT_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CATEGORY_INDICATOR_INDICATOR_TYPE_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await user.type(screen.getByLabelText(/Weight/), '1')
    expect(getByTestId(CATEGORY_INDICATOR_WEIGHT_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)

    await act(() =>
      fireEvent.keyDown(
        getByTestId(CATEGORY_INDICATOR_INDICATOR_TYPE_TEST_ID).childNodes[1],
        { key: 'ArrowDown' }
      )
    )

    await waitFor(() => {
      fireEvent.click(getByText(CATEGORY_INDICATOR_INDICATOR_TYPE_OPTION_LABEL))
      expect(getByText(CATEGORY_INDICATOR_INDICATOR_TYPE_OPTION_LABEL)).toBeInTheDocument()
    })

    await act(() => fireEvent.submit(getByTestId(SUBMIT_BUTTON_TEST_ID)))
    expect(getByTestId(CATEGORY_INDICATOR_NAME_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CATEGORY_INDICATOR_DESCRIPTION_TEST_ID)).toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CATEGORY_INDICATOR_WEIGHT_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(getByTestId(CATEGORY_INDICATOR_INDICATOR_TYPE_TEST_ID)).not.toHaveTextContent(REQUIRED_FIELD_MESSAGE)
    expect(container).toMatchSnapshot()
  })

  describe('Should display toast on submit -', () => {
    test('Success.', async () => {
      mockNextAuthUseSession(statuses.AUTHENTICATED, { isAdminUser: true })
      const mockCreateCategoryIndicator = generateMockApolloData(
        CREATE_CATEGORY_INDICATOR,
        {
          slug: 'test_category_indicator',
          name: 'Test Category Indicator',
          rubricCategorySlug: 'rubric_category_slug',
          weight: 1,
          indicatorType: 'numeric',
          dataSource: 'Test data source',
          scriptName: 'Test script name',
          description: 'Test description'
        },
        null,
        createCategoryIndicatorSuccess
      )
      const { container, getByTestId } = render(
        <CustomMockedProvider mocks={[mockCreateCategoryIndicator]} addTypename={false}>
          <CategoryIndicatorForm categoryIndicator={categoryIndicator} rubricCategory={rubricCategory} />
        </CustomMockedProvider>
      )

      await waitForAllEffectsAndSelectToLoad(container)
      await act(() => fireEvent.submit(getByTestId(SUBMIT_BUTTON_TEST_ID)))
      expect(await screen.findByText('Category Indicator submitted successfully')).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })

    test('Failure.', async () => {
      mockNextAuthUseSession(statuses.AUTHENTICATED, { isAdminUser: true })
      const mockCreateCategoryIndicator = generateMockApolloData(
        CREATE_CATEGORY_INDICATOR,
        {
          slug: 'test_category_indicator',
          name: 'Test Category Indicator',
          weight: 1,
          indicatorType: 'numeric',
          dataSource: 'Test data source',
          scriptName: 'Test script name',
          description: 'Test description'
        },
        createCategoryIndicatorFailure
      )
      const { container, getByTestId } = render(
        <CustomMockedProvider mocks={[mockCreateCategoryIndicator]}>
          <CategoryIndicatorForm categoryIndicator={categoryIndicator} />
        </CustomMockedProvider>
      )

      await waitForAllEffectsAndSelectToLoad(container)
      await act(() => fireEvent.submit(getByTestId(SUBMIT_BUTTON_TEST_ID)))
      expect(await screen.findByText('Category Indicator submission failed')).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })
  })
})
