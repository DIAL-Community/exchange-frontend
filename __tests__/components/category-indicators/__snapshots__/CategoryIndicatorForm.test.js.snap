// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for CategoryIndicatorForm component. Should render CategoryIndicatorForm component for admin user. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="flex flex-col"
  >
    <div
      class="hidden lg:block px-8"
    >
      <div
        class="bg-white pb-3 lg:py-4 whitespace-nowrap text-ellipsis overflow-hidden"
      >
        <a
          class="inline text-dial-blue h5"
          href="/"
        >
          Home
        </a>
        <div
          class="inline h5"
        >
           &gt; 
          <a
            class="text-dial-gray-dark"
            href="/"
          />
        </div>
      </div>
    </div>
    <div
      class="pb-8 px-8"
    >
      <div
        class="sm:px-0 max-w-full mx-auto"
        id="content"
      >
        <form>
          <div
            class="bg-edit shadow-md rounded px-8 pt-6 pb-12 mb-4 flex flex-col gap-3"
          >
            <div
              class="text-2xl font-bold text-dial-blue pb-4"
            >
              Create New Category Indicator
            </div>
            <div
              class="flex flex-col lg:flex-row gap-4"
            >
              <div
                class="w-full lg:w-1/2 flex flex-col gap-y-3"
              >
                <div
                  class="form-field-wrapper"
                  data-testid="category-indicator-name"
                >
                  <label
                    class="form-field-label required-field"
                    for="name"
                  >
                    Name
                  </label>
                  <input
                    class="w-full"
                    id="name"
                    name="name"
                    placeholder="Name"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="form-field-wrapper"
                  data-testid="category-indicator-indicator-type"
                >
                  <label
                    class="form-field-label required-field"
                  >
                    Indicator type
                  </label>
                  <div
                    class=" css-l45kjq-container"
                  >
                    <span
                      class="css-1f43avz-a11yText-A11yText"
                      id="react-select-:r0:-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class="react-select__control css-qv4grc-control"
                    >
                      <div
                        class="react-select__value-container css-17t7gkf-ValueContainer"
                      >
                        <div
                          class="react-select__placeholder css-5rjclb-placeholder"
                          id="react-select-:r0:-placeholder"
                        >
                          Indicator type
                        </div>
                        <div
                          class="react-select__input-container css-qbdosj-Input"
                          data-value=""
                        >
                          <input
                            aria-autocomplete="list"
                            aria-describedby="react-select-:r0:-placeholder"
                            aria-expanded="false"
                            aria-haspopup="true"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            class="react-select__input"
                            id="react-select-id-indicatorType"
                            role="combobox"
                            spellcheck="false"
                            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                      <div
                        class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
                      >
                        <span
                          class="react-select__indicator-separator css-nyxz9h-indicatorSeparator"
                        />
                        <div
                          aria-hidden="true"
                          class="react-select__indicator react-select__dropdown-indicator css-2hgo35-indicatorContainer"
                        >
                          <svg
                            aria-hidden="true"
                            class="text-2xl"
                            fill="none"
                            height="1em"
                            stroke="currentColor"
                            stroke-width="2"
                            viewBox="0 0 24 24"
                            width="1em"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
                              stroke-linecap="round"
                              stroke-linejoin="round"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="form-field-wrapper"
                  data-testid="category-indicator-weight"
                >
                  <label
                    class="form-field-label required-field"
                    for="weight"
                  >
                    Weight
                  </label>
                  <input
                    class="w-full"
                    id="weight"
                    name="weight"
                    placeholder="Weight"
                    step="0.001"
                    type="number"
                    value=""
                  />
                </div>
                <div
                  class="form-field-wrapper"
                >
                  <label
                    class="form-field-label"
                  >
                    Data source
                  </label>
                  <input
                    class="w-full"
                    name="dataSource"
                    placeholder="Data source"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="form-field-wrapper"
                >
                  <label
                    class="form-field-label"
                  >
                    Script name
                  </label>
                  <input
                    class="w-full"
                    name="scriptName"
                    placeholder="Script name"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="w-full lg:w-1/2"
              >
                <div
                  class="flex flex-col gap-y-2"
                  data-testid="category-indicator-description"
                >
                  <label
                    class="form-field-label required-field"
                  >
                    Description
                  </label>
                  <div
                    class="relative"
                  >
                    <svg
                      class="absolute text-lg inset-x-1/2 top-10 spinner"
                      fill="currentColor"
                      height="2em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="2em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
                      />
                    </svg>
                    <div
                      class="html-editor"
                    >
                      <textarea
                        id="description-editor"
                        style="visibility: hidden;"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex flex-wrap text-xl mt-8 gap-3"
            >
              <button
                class="submit-button"
                data-testid="submit-button"
                type="submit"
              >
                Submit Category Indicator
              </button>
              <button
                class="cancel-button"
                type="button"
              >
                Cancel
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
